File: citizensdk/android/src/main/kotlin/org/citizen/sdk/CitizenSdkFlutterWalletFlow.kt
package org.citizen.sdk

import androidx.fragment.app.FragmentActivity
import org.citizen.sdk.ui.CitizenSdkWalletFlowContract
import org.citizen.sdk.ui.CitizenSdkWalletFlowCoordinator
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

/** One-shot ownership registry whose callback may win before handle binding. */
internal class CitizenSdkFlutterOneShotRegistry<K : Any, H : Any>(
    private val cancelHandle: (H) -> Unit,
) {
    internal class Entry<H : Any> {
        val handle = AtomicReference<H?>()
        val finished = AtomicBoolean(false)
        val cancelRequested = AtomicBoolean(false)
    }

    private val entries = ConcurrentHashMap<K, Entry<H>>()

    fun reserve(key: K): Entry<H>? = Entry<H>().takeIf { entries.putIfAbsent(key, it) == null }

    fun bind(entry: Entry<H>, handle: H) {
        check(entry.handle.compareAndSet(null, handle))
        if (entry.cancelRequested.get()) cancelHandle(handle)
    }

    fun finish(key: K, entry: Entry<H>): Boolean {
        if (!entry.finished.compareAndSet(false, true)) return false
        return entries.remove(key, entry)
    }

    fun cancelWhere(predicate: (K) -> Boolean) {
        entries.entries.filter { predicate(it.key) }.forEach { (_, entry) ->
            if (entry.cancelRequested.compareAndSet(false, true)) {
                entry.handle.get()?.let(cancelHandle)
            }
        }
    }

    internal fun sizeForTest(): Int = entries.size
}

/**
 * Secret-free Flutter projection of the SDK-owned wallet Activity.
 *
 * Flutter selects only the operation, word count, or public derivation
 * indices. Recovery phrases and passwords are entered/rendered by the
 * non-exported FLAG_SECURE Activity and never enter this object.
 */
internal class CitizenSdkFlutterWalletFlow {
    private data class Key(val sessionId: String, val requestSequence: Long)
    private val active = CitizenSdkFlutterOneShotRegistry<Key, CitizenSdkWalletFlowCoordinator> {
        it.cancel()
    }

    fun launch(
        sdk: CitizenSdk,
        activity: FragmentActivity?,
        request: CitizenSdkFlutterCodec.Request.SessionRequest,
    ): CompletableFuture<CitizenWalletProfile?> {
        val host = activity ?: return failed(
            CitizenSdkException(
                CitizenSdkErrorCode.UNAVAILABLE,
                "CitizenSDK wallet UI requires a FragmentActivity",
            ),
        )
        val contract = contractRequest(request)
        val key = Key(request.sessionId, request.requestSequence)
        val completion = CompletableFuture<CitizenWalletProfile?>()
        val owner = active.reserve(key) ?: return failed(
            CitizenSdkException(CitizenSdkErrorCode.CONFLICT, "Wallet flow request already exists"),
        )
        try {
            val coordinator = sdk.launchWalletFlow(host, contract) { result ->
                if (!active.finish(key, owner)) return@launchWalletFlow
                when (result) {
                    is CitizenSdkWalletFlowContract.Result.Completed -> completion.complete(result.profile)
                    CitizenSdkWalletFlowContract.Result.Cancelled -> completion.completeExceptionally(
                        CitizenSdkException(CitizenSdkErrorCode.CANCELLED, "CitizenSDK wallet flow cancelled"),
                    )
                    is CitizenSdkWalletFlowContract.Result.Failed -> completion.completeExceptionally(result.error)
                }
            }
            active.bind(owner, coordinator)
        } catch (error: Throwable) {
            if (active.finish(key, owner)) completion.completeExceptionally(error)
        }
        return completion
    }

    /** Cancels every SDK-owned UI flow before supervised session destruction. */
    fun cancelSession(sessionId: String) {
        active.cancelWhere { it.sessionId == sessionId }
    }

    companion object {
        internal fun contractRequest(
            request: CitizenSdkFlutterCodec.Request.SessionRequest,
        ): CitizenSdkWalletFlowContract.Request = when (request) {
            is CitizenSdkFlutterCodec.Request.CreateWallet ->
                CitizenSdkWalletFlowContract.Request.Create(request.wordCount)
            is CitizenSdkFlutterCodec.Request.AddWalletAccounts ->
                CitizenSdkWalletFlowContract.Request.AddAccounts(request.indices)
            is CitizenSdkFlutterCodec.Request.Empty -> {
                require(request.method == "importWallet")
                CitizenSdkWalletFlowContract.Request.Import()
            }
            else -> throw CitizenSdkException(
                CitizenSdkErrorCode.INVALID_ARGUMENT,
                "Request is not an SDK-owned wallet flow",
            )
        }

        private fun <T> failed(error: Throwable): CompletableFuture<T> =
            CompletableFuture<T>().also { it.completeExceptionally(error) }
    }
}

File: citizensdk/darwin/Sources/CitizenSDKFlutter/CitizenSdkFlutterWalletFlow.swift
import CitizenSDK
import Foundation

#if os(iOS)
import UIKit
#elseif os(macOS)
import AppKit
#endif

/// Secret-free bridge to the SDK-owned Apple wallet interface.
@MainActor
internal final class CitizenSdkFlutterWalletFlow {
    private struct Key: Hashable { let session: String; let sequence: Int64 }
    private final class Entry {
        var handle: CitizenSDKWalletFlow?
        var cancelRequested = false
        var finished = false
    }
    private var active: [Key: Entry] = [:]

    func launch(sdk: CitizenSDK, request: CitizenSdkFlutterCodec.Request) async throws -> CitizenWalletProfile? {
        guard let session = request.sessionID, let sequence = request.sequence else {
            throw CitizenSDKError(.invalidArgument, "wallet flow requires a session request")
        }
        let contract = try Self.contract(for: request)
        let key = Key(session: session, sequence: sequence)
        guard active[key] == nil else { throw CitizenSDKError(.conflict, "wallet flow request already exists") }
        let entry = Entry(); active[key] = entry
        return try await withCheckedThrowingContinuation { continuation in
            do {
                #if os(iOS)
                guard let presenter = Self.presenter() else { throw CitizenSDKError(.unavailable, "CitizenSDK wallet UI has no foreground presenter") }
                let handle = try sdk.presentWalletFlow(from: presenter, request: contract) { [weak self, weak entry] result in
                    Task { @MainActor in self?.finish(key, entry: entry, result: result, continuation: continuation) }
                }
                #elseif os(macOS)
                guard let window = NSApp.keyWindow ?? NSApp.mainWindow else { throw CitizenSDKError(.unavailable, "CitizenSDK wallet UI has no host window") }
                let handle = try sdk.presentWalletFlow(from: window, request: contract) { [weak self, weak entry] result in
                    Task { @MainActor in self?.finish(key, entry: entry, result: result, continuation: continuation) }
                }
                #endif
                entry.handle = handle
                if entry.cancelRequested { handle.cancel() }
            } catch {
                active.removeValue(forKey: key)
                continuation.resume(throwing: error)
            }
        }
    }

    static func contract(for request: CitizenSdkFlutterCodec.Request) throws -> CitizenSDKWalletFlowRequest {
        switch request {
        case let .create(_, _, words): return .create(wordCount: words)
        case let .empty(method, _, _) where method == "importWallet": return .importWallet
        case let .addAccounts(_, _, indices): return .addAccounts(indices: indices)
        default: throw CitizenSDKError(.invalidArgument, "request is not an SDK-owned wallet flow")
        }
    }

    func cancelSession(_ session: String) {
        active.filter { $0.key.session == session }.forEach { _, entry in
            entry.cancelRequested = true
            entry.handle?.cancel()
        }
    }

    private func finish(_ key: Key, entry: Entry?, result: CitizenSDKWalletFlowResult,
                        continuation: CheckedContinuation<CitizenWalletProfile?, Error>) {
        guard let entry, active[key] === entry, !entry.finished else { return }
        entry.finished = true; active.removeValue(forKey: key)
        switch result {
        case let .completed(profile): continuation.resume(returning: profile)
        case .cancelled: continuation.resume(throwing: CitizenSDKError(.cancelled, "CitizenSDK wallet flow cancelled"))
        case let .failed(error): continuation.resume(throwing: error)
        @unknown default:
            continuation.resume(throwing: CitizenSDKError(.integrity, "CitizenSDK wallet flow returned an unknown result"))
        }
    }

    #if os(iOS)
    private static func presenter() -> UIViewController? {
        let root = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first(where: \.isKeyWindow)?.rootViewController
        var current = root
        while let presented = current?.presentedViewController { current = presented }
        if let navigation = current as? UINavigationController { return navigation.visibleViewController ?? navigation }
        if let tabs = current as? UITabBarController { return tabs.selectedViewController ?? tabs }
        return current
    }
    #endif
}
